feat: add grpc to glossary

Sonichigo opened the pull request:
### Describe your changes
This pull request adds a new conceptual documentation page for gRPC, covering its purpose, benefits, and use cases in modern distributed systems and microservices architectures.
* Added `content/en/grpc.md` to provide an overview of gRPC, including its architecture (HTTP/2, Protocol Buffers), the problems it solves in distributed systems, and its advantages for microservices, streaming, and interoperability.

### Related issue number or link (ex: `resolves #issue-number`)
resolves #1038 

### Checklist before opening this PR (put `x` in the checkboxes)
- [x] This PR does not contain plagiarism
  - don’t copy other people’s work unless you are quoting and contributing it to them.
- [x] I have signed off on all commits 
  - [signing off](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) (ex: `git commit -s`) is to affirm that commits comply [DCO](https://wiki.linuxfoundation.org/dco). If you are working locally, you could add an alias to your `gitconfig` by running `git config --global alias.ci "commit -s"`.

Sonichigo committed:
feat: add grpc to glossary

Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

chalin reviewed (commented):
Nice idea to add gRPC.
Can someone from the gRPC team review this? @gnossen @ejona86 @thisisnotapril

See inline comments.

/cc @nate-double-u

Sonichigo committed:
Update content/en/grpc.md

Co-authored-by: Patrice Chalin <chalin@users.noreply.github.com>
Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
add link to gRPC website in documentation

Added a link to the gRPC website for reference.

Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo commented:
@iamNoah1 @nate-double-u can i get review?

iamNoah1 reviewed (changes requested):
@Sonichigo Thanks a lot for tackling this term and the effort you put in it. While it might be technically accurate I find it very complex to read. It references a lot of other technical terms, making it hard to understand for someone who is maybe not so much a technical person. Even if, it is not easy. 
My encouragement is to double check on our styleguide: https://glossary.cncf.io/style-guide/ to understand how we want terms to be described ideally. 
Please also pay attention to use semantic line breaks.

hlipsig reviewed (commented):
Agreed with  the previous review. Fantastic idea to include gPRC and an excellent effort. However, I also agree it's diffficult to understand without more of an anchor. I suspect this article could benefit from a comparison to help readers understand the why of gRPC.  For example what the alternatives are and a more concise example of a use case in the "How it helps" section.

jihoon-seo commented on content/en/grpc.md:
```diff
@@ -0,0 +1,27 @@
+
+## Problem it solves?
+
+gRPC addresses common [distributed-systems](./distributed-systems.md) challenges such as inefficient text-based payloads, inconsistent service contracts across languages, limited streaming support, and high-latency multiplexing over HTTP/1.1. 
```
```suggestion
gRPC addresses common [distributed-systems](/distributed-systems/) challenges such as inefficient text-based payloads, inconsistent service contracts across languages, limited streaming support, and high-latency multiplexing over HTTP/1.1. 
```

jihoon-seo commented on content/en/grpc.md:
```diff
@@ -0,0 +1,27 @@
+
+gRPC addresses common [distributed-systems](./distributed-systems.md) challenges such as inefficient text-based payloads, inconsistent service contracts across languages, limited streaming support, and high-latency multiplexing over HTTP/1.1. 
+It also reduces boilerplate by generating code from a single source of truth.
+In modern [microservices architectures](./microservices-architecture.md), these issues become critical as services frequently communicate with each other. 
```
```suggestion
In modern [microservices architectures](/microservices-architecture/), these issues become critical as services frequently communicate with each other. 
```

jihoon-seo commented on content/en/grpc.md:
```diff
@@ -0,0 +1,27 @@
+
+For **microservices**, gRPC simplifies service-to-service communication and scales well in polyglot environments. 
+Its **bi-directional streaming** makes it suitable not just for request/response patterns but also for real-time event-driven use cases such as log aggregation, IoT telemetry, or live data pipelines. 
+While it is not a replacement for [event streaming](./event-streaming.md) platforms like Kafka, gRPC often complements them by providing fast, low-latency RPC communication at the service layer.  
```
```suggestion
While it is not a replacement for [event streaming](/event-streaming/) platforms like Kafka, gRPC often complements them by providing fast, low-latency RPC communication at the service layer.  
```

jihoon-seo reviewed (commented):
The links are not working as-is.

Sonichigo committed:
Update content/en/grpc.md

Co-authored-by: Jihoon Seo <46767780+jihoon-seo@users.noreply.github.com>
Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
Update content/en/grpc.md

Co-authored-by: Jihoon Seo <46767780+jihoon-seo@users.noreply.github.com>
Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
Update content/en/grpc.md

Co-authored-by: Jihoon Seo <46767780+jihoon-seo@users.noreply.github.com>
Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
updates based on feedback

Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
updates based on feedback

Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

nate-double-u commented on content/en/grpc.md:
```diff
@@ -0,0 +1,49 @@
+
+**Simple Example**: Consider a ride-sharing app where the mobile app needs to constantly track driver locations and receive real-time updates. With gRPC's bi-directional streaming, the driver app can continuously send location updates while the server pushes route changes and passenger information back—all efficiently and with minimal latency.
+
+While gRPC is great for service-to-service communication, it's not meant to replace [event streaming](/event-streaming/) platforms like Kafka for high-volume event processing. Instead, gRPC often complements them by handling fast, direct communication between services.
```
try to avoid contractions
```suggestion
While gRPC is great for service-to-service communication, it is not meant to replace [event streaming](/event-streaming/) platforms like Kafka for high-volume event processing. Instead, gRPC often complements them by handling fast, direct communication between services.
```

Sonichigo committed:
Update content/en/grpc.md

Co-authored-by: Nate W <natew@cncf.io>
Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

Sonichigo committed:
refactor gRPC documentation for clarity

Improves the gRPC concept page to align with documentation definition guidelines and semantic line-break conventions. Based on https://glossary.cncf.io/style-guide/ have fixed the whitespaces and proper line semantic breaks.

Signed-off-by: Animesh Pathak <53110238+Sonichigo@users.noreply.github.com>

nate-double-u reviewed (changes requested):
Hi @Sonichigo, thanks for the updates.

This is quite long for one of our definitions. Please take a look at some of our other work to get a sense of how long a definition should be.

This still doesn't follow [our style guide](https://glossary.cncf.io/style-guide/):
- For instance, the gRPC vs. Alternatives section doesn't show up in any of our other work and should be omitted here. 
- Things like contractions should be avoided, generally. You can also find a list of tags there (the tags in this definition don't quite match up).
- In the Problem it addresses section, avoid mentioning the term that is being defined, this section helps folks understand what led to needing the thing.

Sonichigo committed:
feat: refactor grpc definations, fix the spellcheck, resolve the feedback

Signed-off-by: Animesh Pathak <animesh.pathak@harness.io>

nate-double-u commented on content/en/grpc.md:
```diff
@@ -0,0 +1,35 @@
+gRPC solves this by using a compact format to send information, 
+which is faster to transmit and process.
+It also lets you define once how programs should talk to each other, 
+and then automatically creates the necessary code for any programming language you are using.
```
```suggestion
and then automatically creates the necessary code for the programming language you are using.
```

nate-double-u replied:
nit, i don't think gRPC supports all languages yet.

Sonichigo replied:
Right now they support most of them https://grpc.io/docs/languages/.. If needed i can call this out along with maybe add the link to the page?

Sonichigo replied:
@nate-double-u does above look okay?

hlipsig replied:
I'd change it to "and then for various programming languages the protocol supports automatically creates the necessary code."

Sonichigo replied:
Thank @hlipsig for the suggestion, making the changes.

nate-double-u reviewed (commented):
Thanks for the update @Sonichigo, this looks much better to me.

@ejona86, @chalin could you take another look to make sure it's still technically correct? (and sorry for the review delays)

hlipsig commented on content/en/grpc.md:
```diff
@@ -0,0 +1,35 @@
+tags: ["networking", ""]
+---
+
+[gRPC](https://grpc.io/) is a way for different programs to talk to each other over a network.
```
gRPC is an initialism. I think it would be good to state what it stands for.

Sonichigo replied:
@hlipsig i have updated the doc based on feedback

Sonichigo committed:
feat: incorporate the feedback and update the doc

Signed-off-by: Animesh Pathak <animesh.pathak@harness.io>

hlipsig reviewed (approved):
LGTM

nate-double-u commented on content/en/grpc.md:
```diff
@@ -0,0 +1,35 @@
+tags: ["networking", ""]
+---
+
+Google Remote Procedure Calls or [gRPC](https://grpc.io/) is a way for different programs to talk to each other over a network.
```
The acronym reference here is incorrect: https://grpc.io/docs/what-is-grpc/faq/#what-does-grpc-stand-for

gRPC is a recursive acronym, so the "g" doesn't refer to Google.

```suggestion
[gRPC](https://grpc.io/) is a way for different programs to talk to each other over a network.
```

nate-double-u commented on content/en/grpc.md:
```diff
@@ -0,0 +1,35 @@
+like having a conversation instead of exchanging letters.
+For example, in a ride-sharing app, 
+a driver's phone can continuously send location updates 
+while receiving new passenger requests and route changes at the same time.
```
```suggestion
while receiving new passenger requests and route changes.
```

nate-double-u reviewed (dismissed):
Just a couple more changes, I think, before we merge this in:

nate-double-u committed:
Update content/en/grpc.md

Signed-off-by: Nate W <natew@cncf.io>

nate-double-u committed:
Update content/en/grpc.md

Signed-off-by: Nate W <natew@cncf.io>

nate-double-u reviewed (approved):
Thanks, everyone! I think we can merge this in as is now.
